curta/src/chip/hash/keccak256/mod.rs

```diff
@@ -0,0 +1,468 @@
+                        operations,
+                    );
+                    self.set_to_expression(&state_after_xor.get(idx), tmp.expr());
+                } else {
+                    // c bits are directly copied over since it's never changed.
+                    self.set_to_expression(
+                        &state_after_xor.get(idx),
+                        state_after_iota.get(idx).expr(),
+                    );
+                }
+            }
+        }
+
+        // set next state input to keccak_f
+        // if cycle not end, i.e cycle_bit = 0 (within same block)
+        // set to state_after_iota
+        // else
+        //    if still have msg block to process, i.e msg_bit = 0
+        //        set to state_after_xor
+        //    else
+        //        set to initial state, i.e. msg_block's value (a xor 0 = a and c bit are all zero)
+        for x in 0..5 {
+            for y in 0..5 {
+                let idx = x + y * 5;
+                self.set_to_expression_transition(
+                    &state.get(idx).next(),
+                    state_after_iota.get(idx).expr() * cycle_24_end_bit.not_expr()
+                        + (state_after_xor.get(idx).expr() * msg_end_bit.not_expr()
+                            + msg_block.get(idx).expr() * msg_end_bit.expr())
+                            * cycle_24_end_bit.expr(),
+                );
+            }
+        }
+
+        Keccak256Gadget {
+            state,
+            round_constant,
+            cycle_24_end_bit,
+            msg_end_bit,
+            msg_block,
+        }
+    }
+}
+
+impl Keccak256Gadget {
+    pub fn generate_trace<F: Field, I: IntoIterator>(
+        &self,
+        _padded_messages: I,
+        _writer: &TraceWriter<F>,
+    ) -> Keccak256PublicData
+    where
+        I::Item: Borrow<[u8]>,
+    {
+        // TODO: generate the trace so it can satisfy the constraint,
+        // note it's tricky since there're many rows where only one column (set of column) is changing, all the other column including state register are basically directly copying to next row, need to carefully compute the gap in the table and accurately put the right value in right row (position), is it correct?....
+        // the other cleaner method would be instead of expand to other register(column), use same state register but is the .next() in-place, if not, how to constrain multiple rows, use .next().next() ..? is it clean?
+        // currently there's indeed too many columns, which is bad...for both correct trace generation implementation and performance (proof generation time) i think..
+        // also figure out how to pad and discard for variable length input, discard may need some bit array tricks.
+        todo!()
+    }
+
+    // pad for keccak256, msg is array of bytes, but return it's series of bits, i.e. 0 and 1, but just represented as array of u8.
+    // if this confusing may change to some other name and structure
```

Comment:
You can also assume the message will be in bytes so that padding can happen in bytes as well.